libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+{
+	UNUSED_PARAMETER(c);
+	UNUSED_PARAMETER(userdata);
+
+	blog(LOG_INFO, "Server name: '%s %s'",
+	     i->server_name, i->server_version);
+
+	pulseaudio_signal(0);
+}
+
+static void
+pulseaudio_source_info(pa_context *c, const pa_source_info *i, int eol,
+                       void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	PULSE_DATA(userdata);
+	// An error occured
+	if (eol < 0) {
+		data->format = PA_SAMPLE_INVALID;
+		goto skip;
+	}
+	// Terminating call for multi instance callbacks
+	if (eol > 0)
+		goto skip;
+
+	blog(LOG_INFO, "Audio format: %s, %"
+		PRIu32
+		" Hz"
+			", %"
+		PRIu8
+		" channels",
+	     pa_sample_format_to_string(i->sample_spec.format),
+	     i->sample_spec.rate,
+	     i->sample_spec.channels);
+
+	pa_sample_format_t format = i->sample_spec.format;
+	if (pulseaudio_to_obs_audio_format(format) == AUDIO_FORMAT_UNKNOWN) {
+		format = PA_SAMPLE_S16LE;
+
+		blog(LOG_INFO, "Sample format %s not supported by OBS,"
+			"using %s instead for recording",
+		     pa_sample_format_to_string(i->sample_spec.format),
+		     pa_sample_format_to_string(format));
+	}
+
+	uint8_t channels = i->sample_spec.channels;
+	if (pulseaudio_channels_to_obs_speakers(channels) == SPEAKERS_UNKNOWN) {
+		channels = 2;
+
+		blog(LOG_INFO, "%c channels not supported by OBS,"
+			"using %c instead for recording",
+		     i->sample_spec.channels,
+		     channels);
+	}
+
+	data->format = format;
+	data->samples_per_sec = i->sample_spec.rate;
+	data->channels = channels;
+skip:
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_stop_playback(struct audio_monitor *monitor)
+{
+	if (monitor->stream) {
+		pa_stream_disconnect(monitor->stream);
+		pa_stream_unref(monitor->stream);
+		monitor->stream = NULL;
+	}
+
+	blog(LOG_INFO, "Stopped Monitoring in '%s'", monitor->device);
+	blog(LOG_INFO, "Got %"
+		PRIuFAST32
+		" packets with %"
+		PRIuFAST64
+		" frames",
+	     monitor->packets, monitor->frames);
+
+	monitor->packets = 0;
+	monitor->frames = 0;
+}
+
+static bool audio_monitor_init(struct audio_monitor *monitor,
+                               obs_source_t *source)
+{
+	pthread_mutex_init_value(&monitor->playback_mutex);
+
+	monitor->source = source;
+
+	const char *id = obs->audio.monitoring_device_id;
+	if (!id) {
+		return false;
+	}
+
+	if (source->info.output_flags & OBS_SOURCE_DO_NOT_SELF_MONITOR) {
+		obs_data_t *s = obs_source_get_settings(source);
+		const char *s_dev_id = obs_data_get_string(s, "device_id");
+		bool match = devices_match(s_dev_id, id);
+		obs_data_release(s);
+
+		if (match) {
+			monitor->ignore = true;
+			blog(LOG_INFO, "Prevented feedback-loop in '%s'",
+			     s_dev_id);
+			return true;
+		}
+	}
+
+	pulseaudio_init();
+
+	if (strcmp(id, "default") == 0) {
+		get_default_id(&monitor->device);
+	} else {
+		monitor->device = bstrdup(id);
+	}
+
+	if (!monitor->device)
+		return false;
+
+	if (pulseaudio_get_server_info(pulseaudio_server_info,
+	                               (void *) monitor) < 0) {
```

Comment:
Style: Indent with one tab for continuing a control structure or loop statement (if, while, for, etc.).